### PR TITLE
[TECHNICAL] Unit tests for repository classes - Part 1

### PR DESCRIPTION
**Pull request by Aitorbp**
## Related Issues
App: https://github.com/owncloud/android/issues/4232

- [x] Added changelog files for the fixed issues in folder changelog/unreleased. More info [here](https://github.com/owncloud/android/tree/master/changelog#create-changelog-items)
_____

### CODE REVIEW

**Review by JuancaG05 — CHANGES_REQUESTED** [review]
Several changes requested here @Aitorbp, you should pay more attention to previous reviews 😅
Also, in this case calens is not working because the name of the branch is incorrect. Pay attention to these details, they can break our normal workflow

File: owncloudData/src/test/java/com/owncloud/android/data/appregistry/repository/OCAppRegistryRepositoryTest.kt [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+import com.owncloud.android.testutil.OC_FILE
+import io.mockk.every
+import io.mockk.mockk
+import io.mockk.verify
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.flow.flowOf
+import kotlinx.coroutines.test.runTest
+import org.junit.Test
+
+import org.junit.Assert.*
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/appregistry/repository/OCAppRegistryRepositoryTest.kt:40]: Done

File: owncloudData/src/test/java/com/owncloud/android/data/appregistry/repository/OCAppRegistryRepositoryTest.kt [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.owncloud.android.data.appregistry.repository
+
+import com.owncloud.android.data.appregistry.OCAppRegistryRepository
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/appregistry/repository/OCAppRegistryRepositoryTest.kt:23]: done

File: owncloudData/src/test/java/com/owncloud/android/data/appregistry/repository/OCAppRegistryRepositoryTest.kt [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+import kotlinx.coroutines.flow.flowOf
+import kotlinx.coroutines.test.runTest
+import org.junit.Test
+
+import org.junit.Assert.*
+
+class OCAppRegistryRepositoryTest {
+
+    private val localAppRegistryDataSource = mockk<LocalAppRegistryDataSource>(relaxUnitFun = true)
+    private val remoteAppRegistryDataSource = mockk<RemoteAppRegistryDataSource>(relaxUnitFun = true)
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/appregistry/repository/OCAppRegistryRepositoryTest.kt:45]: Done

File: owncloudData/src/test/java/com/owncloud/android/data/appregistry/repository/OCAppRegistryRepositoryTest.kt [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+    private val localCapabilitiesDataSource = mockk<LocalCapabilitiesDataSource>(relaxUnitFun = true)
+    private val appRegistryRepository = OCAppRegistryRepository(
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/appregistry/repository/OCAppRegistryRepositoryTest.kt:47]: Done

File: owncloudData/src/test/java/com/owncloud/android/data/appregistry/repository/OCAppRegistryRepositoryTest.kt [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+        localAppRegistryDataSource,
+        remoteAppRegistryDataSource,
+        localCapabilitiesDataSource,
+    )
+
+    @Test
+    fun `refreshAppRegistryForAccount get the AppRegistry object relate to an account and save it`() {
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/appregistry/repository/OCAppRegistryRepositoryTest.kt:54]: Done

File: owncloudData/src/test/java/com/owncloud/android/data/appregistry/repository/OCAppRegistryRepositoryTest.kt [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+
+        every { localCapabilitiesDataSource.getCapabilitiesForAccount(OC_ACCOUNT_NAME) } returns OC_CAPABILITY
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/appregistry/repository/OCAppRegistryRepositoryTest.kt:56]: Done

File: owncloudTestUtil/src/main/java/com/owncloud/android/testutil/OCCapability.kt [hunk 7]
```diff
@@ -55,3 +54,15 @@ val OC_CAPABILITY =
         filesAppProviders = null,
         spaces = null
     )
+
+val OC_CAPABILITY_WITH_FILE_PROVIDERS = OC_CAPABILITY.copy(
+    accountName = OC_ACCOUNT_NAME,
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudTestUtil/src/main/java/com/owncloud/android/testutil/OCCapability.kt:59]: Done

File: owncloudData/src/test/java/com/owncloud/android/data/appregistry/repository/OCAppRegistryRepositoryTest.kt [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+                flowOf(listOf(OC_APP_REGISTRY_MIMETYPE))
+        val resultActual = appRegistryRepository.getAppRegistryWhichAllowCreation(OC_ACCOUNT_NAME).first()
+
+        assertEquals(listOf(OC_APP_REGISTRY_MIMETYPE), resultActual)
+
+        verify(exactly = 1) { localAppRegistryDataSource.getAppRegistryWhichAllowCreation(OC_ACCOUNT_NAME) }
+    }
+
+    @Test
+    fun `getUrlToOpenInWeb returns a String correctly`() {
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/appregistry/repository/OCAppRegistryRepositoryTest.kt:108]: Done

File: owncloudData/src/test/java/com/owncloud/android/data/appregistry/repository/OCAppRegistryRepositoryTest.kt [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+                openWebEndpoint = openWebEndpoint,
+                fileId = OC_FILE.id.toString(),
+                appName = appName
+            )
+        } returns expectedUrl
+
+        val resultActual = appRegistryRepository.getUrlToOpenInWeb(
+            accountName = OC_ACCOUNT_NAME,
+            openWebEndpoint = openWebEndpoint,
+            fileId = OC_FILE.id.toString(),
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/appregistry/repository/OCAppRegistryRepositoryTest.kt:124]: Done

File: owncloudData/src/test/java/com/owncloud/android/data/appregistry/repository/OCAppRegistryRepositoryTest.kt [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+                openWebEndpoint = openWebEndpoint,
+                fileId = OC_FILE.id.toString(),
+                appName = appName
+            )
+        }
+    }
+
+    @Test
+    fun `createFileWithAppProvider returns a String`() {
+        val createFileWithAppProviderEndpoint = "app/file-with-app-provider-endpoint"
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/appregistry/repository/OCAppRegistryRepositoryTest.kt:142]: Done

File: owncloudData/src/test/java/com/owncloud/android/data/appregistry/repository/OCAppRegistryRepositoryTest.kt [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+
+        assertEquals(listOf(OC_APP_REGISTRY_MIMETYPE), resultActual)
+
+        verify(exactly = 1) { localAppRegistryDataSource.getAppRegistryWhichAllowCreation(OC_ACCOUNT_NAME) }
+    }
+
+    @Test
+    fun `getUrlToOpenInWeb returns a String correctly`() {
+        val expectedUrl = "https://example.com/file123"
+        val openWebEndpoint = "app/open-with-web"
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/appregistry/repository/OCAppRegistryRepositoryTest.kt:110]: Done

File: owncloudData/src/test/java/com/owncloud/android/data/authentication/repository/OCAuthenticationRepositoryTest.kt
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/authentication/repository/OCAuthenticationRepositoryTest.kt:1]: Done

File: owncloudData/src/test/java/com/owncloud/android/data/authentication/repository/OCAuthenticationRepositoryTest.kt [hunk 12]
```diff
@@ -37,285 +36,148 @@ import io.mockk.every
...
 import org.junit.Assert.assertEquals
+import org.junit.Assert.assertTrue
 import org.junit.Test
 
 class OCAuthenticationRepositoryTest {
 
-    private val localAuthenticationDataSource = mockk<LocalAuthenticationDataSource>(relaxed = true)
-    private val remoteAuthenticationDataSource = mockk<RemoteAuthenticationDataSource>(relaxed = true)
+    private val localAuthenticationDataSource = mockk<LocalAuthenticationDataSource>(relaxUnitFun = true)
+    private val remoteAuthenticationDataSource = mockk<RemoteAuthenticationDataSource>(relaxUnitFun = true)
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/authentication/repository/OCAuthenticationRepositoryTest.kt:45]: Done

File: owncloudData/src/test/java/com/owncloud/android/data/authentication/repository/OCAuthenticationRepositoryTest.kt [hunk 13]
```diff
@@ -37,285 +36,148 @@ import io.mockk.every
...
     private val ocAuthenticationRepository: OCAuthenticationRepository =
         OCAuthenticationRepository(localAuthenticationDataSource, remoteAuthenticationDataSource)
 
     @Test
-    fun loginBasicOk() {
-        every { remoteAuthenticationDataSource.loginBasic(any(), any(), any()) } returns Pair(
+    fun `loginBasic returns String with the account name correctly`() {
+        every {
+            remoteAuthenticationDataSource.loginBasic(
+                serverPath = any(),
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/authentication/repository/OCAuthenticationRepositoryTest.kt:53]: Done

File: owncloudData/src/test/java/com/owncloud/android/data/authentication/repository/OCAuthenticationRepositoryTest.kt [hunk 14]
```diff
@@ -37,285 +36,148 @@ import io.mockk.every
...
     }
 
     @Test
-    fun loginBasicOAuth() {
-        every { remoteAuthenticationDataSource.loginOAuth(any(), any(), any()) } returns Pair(
-            OC_USER_INFO,
-            OC_REDIRECTION_PATH.lastPermanentLocation
-        )
-
+    fun `loginOAuth returns String with the account name correctly`() {
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/authentication/repository/OCAuthenticationRepositoryTest.kt:96]: done

File: owncloudData/src/test/java/com/owncloud/android/data/authentication/repository/OCAuthenticationRepositoryTest.kt [hunk 15]
```diff
@@ -37,285 +36,148 @@ import io.mockk.every
...
-                any(),
-                any(),
-                any(),
-                any(),
-                any(),
-                any(),
-                any(),
-                any()
+            remoteAuthenticationDataSource.loginOAuth(
+                serverPath = any(),
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/authentication/repository/OCAuthenticationRepositoryTest.kt:99]: Done

File: owncloudData/src/test/java/com/owncloud/android/data/authentication/repository/OCAuthenticationRepositoryTest.kt [hunk 16]
```diff
@@ -37,285 +36,148 @@ import io.mockk.every
...
+                username = OC_BASIC_USERNAME,
+                accessToken = any()
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/authentication/repository/OCAuthenticationRepositoryTest.kt:101]: Done

File: owncloudData/src/test/java/com/owncloud/android/data/authentication/repository/OCAuthenticationRepositoryTest.kt [hunk 17]
```diff
@@ -37,285 +36,148 @@ import io.mockk.every
...
-                any()
+                userName = OC_BASIC_USERNAME,
+                lastPermanentLocation = OC_REDIRECTION_PATH.lastPermanentLocation,
+                authTokenType = OC_AUTH_TOKEN_TYPE,
+                accessToken = OC_ACCESS_TOKEN,
+                serverInfo = OC_SECURE_SERVER_INFO_BASIC_AUTH,
+                userInfo = OC_USER_INFO,
+                refreshToken = OC_REFRESH_TOKEN,
+                scope = OC_SCOPE,
+                updateAccountWithUsername = any(),
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/authentication/repository/OCAuthenticationRepositoryTest.kt:118]: Done

File: owncloudData/src/test/java/com/owncloud/android/data/authentication/repository/OCAuthenticationRepositoryTest.kt [hunk 18]
```diff
@@ -37,285 +36,148 @@ import io.mockk.every
...
-            localAuthenticationDataSource.supportsOAuth2(any())
+            localAuthenticationDataSource.supportsOAuth2(OC_ACCOUNT_NAME)
         }
     }
 
     @Test
-    fun getBaseUrlOk() {
+    fun `getBaseUrl returns a String with the base url`() {
         every {
             localAuthenticationDataSource.getBaseUrl(any())
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/authentication/repository/OCAuthenticationRepositoryTest.kt:171]: Done

File: owncloudData/src/test/java/com/owncloud/android/data/capabilities/repository/OCCapabilityRepositoryTest.kt [hunk 19]
```diff
@@ -41,46 +42,14 @@ class OCCapabilityRepositoryTest {
     @JvmField
     val instantExecutorRule = InstantTaskExecutorRule()
 
-    private val localCapabilitiesDataSource = mockk<LocalCapabilitiesDataSource>(relaxed = true)
-    private val remoteCapabilitiesDataSource = mockk<RemoteCapabilitiesDataSource>(relaxed = true)
-    private val appRegistryRepository = mockk<AppRegistryRepository>()
+    private val localCapabilitiesDataSource = mockk<LocalCapabilitiesDataSource>(relaxUnitFun = true)
+    private val remoteCapabilitiesDataSource = mockk<RemoteCapabilitiesDataSource>(relaxUnitFun = true)
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/capabilities/repository/OCCapabilityRepositoryTest.kt:46]: Done

File: owncloudData/src/test/java/com/owncloud/android/data/capabilities/repository/OCCapabilityRepositoryTest.kt [hunk 20]
```diff
@@ -41,46 +42,14 @@ class OCCapabilityRepositoryTest {
...
-            remoteCapabilitiesDataSource.getCapabilities(OC_ACCOUNT_NAME)
-        }
-        verify(exactly = 0) {
-            localCapabilitiesDataSource.insertCapabilities(any())
-        }
-    }
-
-    @Test
-    fun getCapabilitiesAsLiveData() {
+    fun `getCapabilitiesAsLiveData returns a livedata of OCCapability correctly`() {
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/capabilities/repository/OCCapabilityRepositoryTest.kt:52]: Done

File: owncloudData/src/test/java/com/owncloud/android/data/capabilities/repository/OCCapabilityRepositoryTest.kt [hunk 21]
```diff
@@ -41,46 +42,14 @@ class OCCapabilityRepositoryTest {
...
         val capabilitiesLiveData = MutableLiveData<OCCapability>()
 
         every {
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/capabilities/repository/OCCapabilityRepositoryTest.kt:55]: Done

File: owncloudData/src/test/java/com/owncloud/android/data/capabilities/repository/OCCapabilityRepositoryTest.kt [hunk 22]
```diff
@@ -96,40 +65,51 @@ class OCCapabilityRepositoryTest {
         capabilitiesToEmit.forEach {
             capabilitiesLiveData.postValue(it)
         }
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/capabilities/repository/OCCapabilityRepositoryTest.kt:67]: Done

File: owncloudData/src/test/java/com/owncloud/android/data/capabilities/repository/OCCapabilityRepositoryTest.kt [hunk 23]
```diff
@@ -96,40 +65,51 @@ class OCCapabilityRepositoryTest {
...
-            localCapabilitiesDataSource.getCapabilitiesForAccountAsLiveData(any())
-        } throws Exception()
-
-        ocCapabilityRepository.getCapabilitiesAsLiveData(OC_ACCOUNT_NAME)
+        verify(exactly = 1) { localCapabilitiesDataSource.getCapabilitiesForAccountAsLiveData(OC_ACCOUNT_NAME) }
     }
 
     @Test
-    fun getStoredCapabilities() {
+    fun `getStoredCapabilities returns an object of OCCapability`() {
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/capabilities/repository/OCCapabilityRepositoryTest.kt:74]: Done

File: owncloudData/src/test/java/com/owncloud/android/data/capabilities/repository/OCCapabilityRepositoryTest.kt [hunk 24]
```diff
@@ -96,40 +65,51 @@ class OCCapabilityRepositoryTest {
...
 
         every {
             localCapabilitiesDataSource.getCapabilitiesForAccount(any())
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/capabilities/repository/OCCapabilityRepositoryTest.kt:77]: Done

File: owncloudData/src/test/java/com/owncloud/android/data/capabilities/repository/OCCapabilityRepositoryTest.kt [hunk 25]
```diff
@@ -96,40 +65,51 @@ class OCCapabilityRepositoryTest {
...
         } returns null
 
         val result = ocCapabilityRepository.getStoredCapabilities(OC_ACCOUNT_NAME)
 
-        Assert.assertEquals(null, result)
+        assertNull(result)
+
+        verify(exactly = 1) { localCapabilitiesDataSource.getCapabilitiesForAccount(OC_ACCOUNT_NAME) }
+
+    }
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/capabilities/repository/OCCapabilityRepositoryTest.kt:100]: Done

File: owncloudData/src/test/java/com/owncloud/android/data/capabilities/repository/OCCapabilityRepositoryTest.kt [hunk 26]
```diff
@@ -96,40 +65,51 @@ class OCCapabilityRepositoryTest {
...
+
+    @Test
+    fun `refreshCapabilitiesForAccount updates capabilities correctly`() {
+
+        every { remoteCapabilitiesDataSource.getCapabilities(any()) } returns OC_CAPABILITY_WITH_FILE_PROVIDERS
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/capabilities/repository/OCCapabilityRepositoryTest.kt:105]: Done

File: owncloudData/src/test/java/com/owncloud/android/data/capabilities/repository/OCCapabilityRepositoryTest.kt [hunk 27]
```diff
@@ -96,40 +65,51 @@ class OCCapabilityRepositoryTest {
...
 
-        Assert.assertEquals(null, result)
+        assertNull(result)
+
+        verify(exactly = 1) { localCapabilitiesDataSource.getCapabilitiesForAccount(OC_ACCOUNT_NAME) }
+
+    }
+
+    @Test
+    fun `refreshCapabilitiesForAccount updates capabilities correctly`() {
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/capabilities/repository/OCCapabilityRepositoryTest.kt:103]: Done

File: owncloudData/src/test/java/com/owncloud/android/data/capabilities/repository/OCCapabilityRepositoryTest.kt
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/capabilities/repository/OCCapabilityRepositoryTest.kt:1]: Done

File: owncloudTestUtil/src/main/java/com/owncloud/android/testutil/OCCapability.kt [hunk 28]
```diff
@@ -55,3 +54,15 @@ val OC_CAPABILITY =
         filesAppProviders = null,
         spaces = null
     )
+
+val OC_CAPABILITY_WITH_FILE_PROVIDERS = OC_CAPABILITY.copy(
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudTestUtil/src/main/java/com/owncloud/android/testutil/OCCapability.kt:58]: Done

File: owncloudTestUtil/src/main/java/com/owncloud/android/testutil/OCCapability.kt
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudTestUtil/src/main/java/com/owncloud/android/testutil/OCCapability.kt:1]: DOne

File: owncloudData/src/test/java/com/owncloud/android/data/appregistry/repository/OCAppRegistryRepositoryTest.kt [hunk 29]
```diff
@@ -0,0 +1,171 @@
...
+                fileId = OC_FILE.id.toString(),
+                appName = appName
+            )
+        }
+    }
+
+    @Test
+    fun `createFileWithAppProvider returns a String`() {
+        val createFileWithAppProviderEndpoint = "app/file-with-app-provider-endpoint"
+        val parentContainerId = OC_FILE.remoteId.toString()
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/appregistry/repository/OCAppRegistryRepositoryTest.kt:143]: Done

File: owncloudData/src/test/java/com/owncloud/android/data/appregistry/repository/OCAppRegistryRepositoryTest.kt [hunk 30]
```diff
@@ -0,0 +1,171 @@
...
+        val fileWithAppProvider = "fileWithAppProvider"
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/appregistry/repository/OCAppRegistryRepositoryTest.kt:144]: Done

File: owncloudData/src/test/java/com/owncloud/android/data/capabilities/repository/OCCapabilityRepositoryTest.kt [hunk 31]
```diff
@@ -41,95 +44,67 @@ class OCCapabilityRepositoryTest {
...
+        every {
+            localCapabilitiesDataSource.getCapabilitiesForAccountAsLiveData(OC_ACCOUNT_NAME)
+        } returns capabilitiesLiveData
 
-    @Test(expected = NoConnectionWithServerException::class)
-    fun refreshCapabilitiesFromNetworkNoConnection() {
+        val capabilitiesEmitted = ocCapabilityRepository.getCapabilitiesAsLiveData(OC_ACCOUNT_NAME).getLastEmittedValue()
 
-        every { remoteCapabilitiesDataSource.getCapabilities(any()) } throws NoConnectionWithServerException()
+        val capabilitiesToEmit = capabilitiesLiveData.getLastEmittedValue()
```
Comment (JuancaG05) [inline owncloudData/src/test/java/com/owncloud/android/data/capabilities/repository/OCCapabilityRepositoryTest.kt:63]: The logical order of a test is first mocking/constructing the expected result and then obtaining the actual result, then comparing both

**Review by JuancaG05 — CHANGES_REQUESTED** [review]
Second round of CR here @Aitorbp

File: owncloudData/src/test/java/com/owncloud/android/data/appregistry/repository/OCAppRegistryRepositoryTest.kt [hunk 32]
```diff
@@ -0,0 +1,171 @@
...
+        }
+    }
+
+    @Test
+    fun `createFileWithAppProvider returns a String with the new file ID`() {
+
+        every {
+            remoteAppRegistryDataSource.createFileWithAppProvider(
+                accountName = OC_ACCOUNT_NAME,
+                createFileWithAppProviderEndpoint = OC_CAPABILITY_WITH_FILES_APP_PROVIDERS.filesAppProviders?.openWebUrl.toString(),
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/appregistry/repository/OCAppRegistryRepositoryTest.kt:148]: Ok, the appropriate one is newUrl.

File: owncloudData/src/test/java/com/owncloud/android/data/capabilities/repository/OCCapabilityRepositoryTest.kt [hunk 33]
```diff
@@ -41,95 +44,67 @@ class OCCapabilityRepositoryTest {
...
-        val capability = OC_CAPABILITY.copy(accountName = OC_ACCOUNT_NAME)
-
-        every { remoteCapabilitiesDataSource.getCapabilities(any()) } returns capability
-
-        ocCapabilityRepository.refreshCapabilitiesForAccount(OC_ACCOUNT_NAME)
-
-        verify(exactly = 1) {
-            remoteCapabilitiesDataSource.getCapabilities(OC_ACCOUNT_NAME)
-        }
+    fun `getCapabilitiesAsLiveData returns a Livedata of OCCapability`() {
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/capabilities/repository/OCCapabilityRepositoryTest.kt:54]: Done

File: owncloudData/src/test/java/com/owncloud/android/data/capabilities/repository/OCCapabilityRepositoryTest.kt (hunk 31 above)
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/capabilities/repository/OCCapabilityRepositoryTest.kt:63]: Changed the order

File: owncloudData/src/test/java/com/owncloud/android/data/appregistry/repository/OCAppRegistryRepositoryTest.kt [hunk 34]
```diff
@@ -0,0 +1,171 @@
...
+import io.mockk.every
+import io.mockk.mockk
+import io.mockk.verify
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.flow.flowOf
+import kotlinx.coroutines.test.runTest
+import org.junit.Assert.assertEquals
+import org.junit.Test
+
+class OCAppRegistryRepositoryTest {
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/appregistry/repository/OCAppRegistryRepositoryTest.kt:41]: Done

File: owncloudData/src/test/java/com/owncloud/android/data/authentication/repository/OCAuthenticationRepositoryTest.kt [hunk 35]
```diff
@@ -37,285 +38,152 @@ import io.mockk.every
...
 import org.junit.Assert.assertEquals
+import org.junit.Assert.assertTrue
 import org.junit.Test
 
 class OCAuthenticationRepositoryTest {
 
-    private val localAuthenticationDataSource = mockk<LocalAuthenticationDataSource>(relaxed = true)
-    private val remoteAuthenticationDataSource = mockk<RemoteAuthenticationDataSource>(relaxed = true)
+    private val localAuthenticationDataSource = mockk<LocalAuthenticationDataSource>()
+    private val remoteAuthenticationDataSource = mockk<RemoteAuthenticationDataSource>(relaxUnitFun = true)
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/authentication/repository/OCAuthenticationRepositoryTest.kt:47]: Done

File: owncloudTestUtil/src/main/java/com/owncloud/android/testutil/OCCapability.kt [hunk 36]
```diff
@@ -55,3 +56,14 @@ val OC_CAPABILITY =
         filesAppProviders = null,
         spaces = null
     )
+
+val OC_CAPABILITY_WITH_FILES_APP_PROVIDERS = OC_CAPABILITY.copy(
+    filesAppProviders = OCCapability.AppProviders(
+        enabled = true,
+        version = "",
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudTestUtil/src/main/java/com/owncloud/android/testutil/OCCapability.kt:63]: Added

File: owncloudData/src/test/java/com/owncloud/android/data/appregistry/repository/OCAppRegistryRepositoryTest.kt [hunk 37]
```diff
@@ -0,0 +1,171 @@
...
+
+    @Test
+    fun `createFileWithAppProvider returns a String with the new file ID`() {
+
+        every {
+            remoteAppRegistryDataSource.createFileWithAppProvider(
+                accountName = OC_ACCOUNT_NAME,
+                createFileWithAppProviderEndpoint = OC_CAPABILITY_WITH_FILES_APP_PROVIDERS.filesAppProviders?.openWebUrl.toString(),
+                parentContainerId = OC_FOLDER.remoteId.toString(),
+                filename = OC_FOLDER.fileName,
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/appregistry/repository/OCAppRegistryRepositoryTest.kt:150]: Done

File: owncloudData/src/test/java/com/owncloud/android/data/appregistry/repository/OCAppRegistryRepositoryTest.kt [hunk 38]
```diff
@@ -0,0 +1,171 @@
...
+    }
+
+    @Test
+    fun `createFileWithAppProvider returns a String with the new file ID`() {
+
+        every {
+            remoteAppRegistryDataSource.createFileWithAppProvider(
+                accountName = OC_ACCOUNT_NAME,
+                createFileWithAppProviderEndpoint = OC_CAPABILITY_WITH_FILES_APP_PROVIDERS.filesAppProviders?.openWebUrl.toString(),
+                parentContainerId = OC_FOLDER.remoteId.toString(),
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/appregistry/repository/OCAppRegistryRepositoryTest.kt:149]: Done

File: owncloudData/src/test/java/com/owncloud/android/data/appregistry/repository/OCAppRegistryRepositoryTest.kt [hunk 39]
```diff
@@ -0,0 +1,171 @@
...
+            appName = appName
+        )
+
+        assertEquals(expectedUrl, resultActual)
+
+        verify(exactly = 1) {
+            remoteAppRegistryDataSource.getUrlToOpenInWeb(
+                accountName = OC_ACCOUNT_NAME,
+                openWebEndpoint = OC_CAPABILITY_WITH_FILES_APP_PROVIDERS.filesAppProviders?.openWebUrl.toString(),
+                fileId = OC_FILE.remoteId.toString(),
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/appregistry/repository/OCAppRegistryRepositoryTest.kt:136]: Done

File: owncloudData/src/test/java/com/owncloud/android/data/appregistry/repository/OCAppRegistryRepositoryTest.kt [hunk 40]
```diff
@@ -0,0 +1,171 @@
...
+                appName = appName
+            )
+        } returns expectedUrl
+
+        val resultActual = ocAppRegistryRepository.getUrlToOpenInWeb(
+            accountName = OC_ACCOUNT_NAME,
+            openWebEndpoint = OC_CAPABILITY_WITH_FILES_APP_PROVIDERS.filesAppProviders?.openWebUrl.toString(),
+            fileId = OC_FILE.remoteId.toString(),
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/appregistry/repository/OCAppRegistryRepositoryTest.kt:126]: Done

File: owncloudData/src/test/java/com/owncloud/android/data/appregistry/repository/OCAppRegistryRepositoryTest.kt [hunk 41]
```diff
@@ -0,0 +1,171 @@
...
+    fun `createFileWithAppProvider returns a String with the new file ID`() {
+
+        every {
+            remoteAppRegistryDataSource.createFileWithAppProvider(
+                accountName = OC_ACCOUNT_NAME,
+                createFileWithAppProviderEndpoint = OC_CAPABILITY_WITH_FILES_APP_PROVIDERS.filesAppProviders?.openWebUrl.toString(),
+                parentContainerId = OC_FOLDER.remoteId.toString(),
+                filename = OC_FOLDER.fileName,
+            )
+        } returns OC_FOLDER.id.toString()
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/appregistry/repository/OCAppRegistryRepositoryTest.kt]: Done

File: owncloudData/src/test/java/com/owncloud/android/data/capabilities/repository/OCCapabilityRepositoryTest.kt [hunk 42]
```diff
@@ -41,95 +44,67 @@ class OCCapabilityRepositoryTest {
...
-
-        every { remoteCapabilitiesDataSource.getCapabilities(any()) } returns capability
-
-        ocCapabilityRepository.refreshCapabilitiesForAccount(OC_ACCOUNT_NAME)
-
-        verify(exactly = 1) {
-            remoteCapabilitiesDataSource.getCapabilities(OC_ACCOUNT_NAME)
-        }
+    fun `getCapabilitiesAsLiveData returns a LiveData of OCCapability`() {
+        val capabilitiesLiveData = MutableLiveData<OCCapability>()
```
Comment (JuancaG05) [inline owncloudData/src/test/java/com/owncloud/android/data/capabilities/repository/OCCapabilityRepositoryTest.kt:55]: Did you realise this is empty? We are then comparing two `null` in the `assertEquals`

**Review by JuancaG05 — CHANGES_REQUESTED** [review]
One more change here @Aitorbp

File: owncloudData/src/test/java/com/owncloud/android/data/capabilities/repository/OCCapabilityRepositoryTest.kt (hunk 42 above)
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/capabilities/repository/OCCapabilityRepositoryTest.kt:55]: Done

**Review by JuancaG05 — APPROVED** [review]
LGTM now!

**Pull request merged**